File: python/task3c/S3_PIPE_01.py
# S3_PIPE_01 - Implement pipe(*fns)

from functools import reduce

def pipe(*fns):
    return lambda x: reduce(lambda v, fn: fn(v), fns, x)

# Test with numeric transforms
add_one = lambda x: x + 1
double = lambda x: x * 2
square = lambda x: x * x

pipeline = pipe(add_one, double, square)

# pipe(add_one, double, square)(3) = square(double(add_one(3)))
# = square(double(4)) = square(8) = 64
print(f"pipe(add_one, double, square)(3) = {pipeline(3)}")
print(f"  3 -> +1 -> 4 -> *2 -> 8 -> ^2 -> 64")

print(f"\npipe(double, add_one)(5) = {pipe(double, add_one)(5)}")

File: js/task3c/S3_PIPE_01.js
// S3_PIPE_01 - pipe()
// left-to-right function composition

const pipe = (...fns) => x => {
  let result = x;
  for (const fn of fns) {
    result = fn(result);
  }
  return result;
};

// tests
const addOne = x => x + 1;
const double = x => x * 2;
const square = x => x * x;

const pipeline = pipe(addOne, double, square);

console.log("pipe(addOne, double, square)(3):");
console.log("  3 -> +1 -> 4 -> *2 -> 8 -> ^2 -> 64");
console.log("  Result:", pipeline(3)); // 64

const another = pipe(double, addOne);
console.log("\npipe(double, addOne)(5):", another(5)); // 11
